iis-wf/iis-wf-ingest-pmc/src/main/java/eu/dnetlib/iis/wf/ingest/pmc/metadata/ArticleMetaXmlHandler.java

```diff
@@ -0,0 +1,234 @@
+        this.builder = builder;
+        this.parents = new Stack<String>();
+    }
+    
+    
+    //------------------------ LOGIC --------------------------
+    
+    @Override
+    public void startDocument() throws SAXException {
+        this.parents = new Stack<String>();
+        builder.setAffiliations(Lists.newArrayList());
+        builder.setAuthors(Lists.newArrayList());
+    }
+    
+    @Override
+    public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
+        
+        
+        if (isElement(qName, ELEM_AFFILIATION)) {
+            affiliationText.setLength(0);
+            currentAffiliationId = attributes.getValue(ATTR_AFFILIATION_ID);
+        } else if (isElement(qName, ELEM_ARTICLE_ID)) {
+            currentArticleIdType = attributes.getValue(PUB_ID_TYPE);
+        } else if (isElement(qName, ELEM_CONTRIBUTOR)) {
+            if (ATTR_VALUE_AUTHOR.equals(attributes.getValue(ATTR_CONTRIBUTOR_TYPE))) {
+                currentAuthor = new PmcIngestAuthor();
+            }
+        } else if (isWithinElement(qName, ELEM_XREF, parents, ELEM_CONTRIBUTOR)) {
+            if (currentAuthor != null && ATTR_AFFILIATION_XREF.equals(attributes.getValue(ATTR_XREF_TYPE))) {
+                String affId = attributes.getValue(ATTR_XREF_ID);
+                if (affId != null) {
+                    currentAuthor.getAffiliationRefId().add(affId);
+                }
+            }
+        }
+        
+        
+        this.parents.push(qName);
+    }
+    
+    @Override
+    public void characters(char[] ch, int start, int length) throws SAXException {
+        
+        this.currentValue = new String(ch, start, length);
+        
+        if (hasAmongParents(parents, ELEM_AFFILIATION)) {
+            
+            // skipping affiliation position element
+            if (!hasAmongParents(parents, ELEM_LABEL) && !hasAmongParents(parents, ELEM_SUP)) {
+                this.affiliationText.append(new String(ch, start, length));
+            }
+            
+        }
+        
+    }
+    
+    @Override
+    public void endElement(String uri, String localName, String qName) throws SAXException {
+        this.parents.pop();
+        
+        if (isElement(qName, ELEM_ARTICLE_ID) && currentArticleIdType != null) {
+            builder.getExternalIdentifiers().put(currentArticleIdType, currentValue.trim());
+        } else if (isElement(qName, ELEM_FPAGE)) {
+            if (builder.getPages()==null) {
+                builder.setPages(Range.newBuilder().build());
+            }
+            builder.getPages().setStart(this.currentValue.trim());
+        } else if (isElement(qName, ELEM_LPAGE)) {
+            if (builder.getPages()==null) {
+                builder.setPages(Range.newBuilder().build());
+            }
+            builder.getPages().setEnd(this.currentValue.trim());
+            
+        } else if (isElement(qName, ELEM_AFFILIATION)) {
```

Comment:
You don't handle affiliations in fn tags, see: http://jats.nlm.nih.gov/archiving/tag-library/1.1d1/n-pjh2.html
I'm not sure if we should though, we'd have to check the data.